### ISSUE TITLE
Need assistance please

### ISSUE THREAD

**Issue by Tomadoggy** [jmc265/node-red-contrib-rpi-433#5]
First of all...thank you for working on this as when I can get it functioning in my node red project...I believe I will finally have an elegant solution!

I am unable to get the rpi-433-sniffer functioning in node red.  (will also need the sender, but am choosing to focus on the listener first).  I have met all of the prerequisite installs and verified that I can see codes on my Pi when using the:
 /home/pi/rpi-rf/scripts/rpi-rf_receive     script.

I have updated my RFSniffer.cpp file to use int pin = 27 (which is the wiring pi pin for my GPIO27 where the 433mhz receiver is installed)  However I am not certain this file is being read when launching rpi-rf_receive as changes to the .cpp file do not change the pin being listened to when launching the receive.

In node red, I simply have the sniffer node deployed with a debug node to see if codes are coming through, but even though using a large variety of alterations within the sniffer node...never had any output to the debug node.  To keep things simple, I am using a single code captured in the rpi-rf_receive that is consistent with a 374 pulselength (however the 433 sender does seem to vary a little in this value with each use)  thus may be somewhat unreliable in actually getting the message to node red as I am assuming that unless the pulselength matches exactly, I would not get a message.  None the less, first things first, can you please help guide me in what I may be missing to actually get the node red sniffer node to see the codes being captured by rpi-rf_receive?

Many thanks in advance

**Comment by jmc265** [comment]
If you are changing the content of RFSniffer.cpp you will need to recompile it before running it again. Also, my version of RFSniffer.cpp does not require you to hard code the PIN within the source code, it uses an argument from the CLI. Please see here: https://github.com/jmc265/rpi-433-v3

I am afraid it is really hard to debug what is going on as I have nothing to work from here. Have you tried different pin numbers as there are a couple of different numbering systems for the pins on the PI? 
On my sniffer node I leave the pulse length set to 0, I don't think that makes too much difference.

**Comment by Tomadoggy** [comment]
Thank you and Happy 2021!  If I don't need to change the .cpp then all the better.  It does make sense that the pin # is being passed from your node, but in reading your documentation and trying to get this to work...I thought I would at least give it a try.  I will leave the .cpp alone unless there is something there you feel I need to change.

I am using the wiring pi (wPi column from gpio readall) and have my 433 receiver on GPIO27 and thus using the corresponding wPi 27.  I am assuming that since pri-rf_receive and seeing the codes that at least the underlying v3 package is functioning properly  (sorry if I am not using the correct term... I am a bit of a NOOB).  And to your question...yes, I have tried every single pin number in the node red sniffer node to see if I received any message, but so far nothing.

I have set the pulse length to 0, but no incoming message from the debug.  I appreciate your quick response, and will willingly provide any debug assistance.  Below is the output on the pi from the receive, and also the node red sniffer setup.  Please let me know what I may do to help move this to success!!

![rpi-rf_receive](https://user-images.githubusercontent.com/62314433/103443577-de2a2a80-4c1d-11eb-848a-ec74990bff07.png)

![NR-sniff](https://user-images.githubusercontent.com/62314433/103443611-4842cf80-4c1e-11eb-8f30-47bbb22c4497.png)

**Comment by jmc265** [comment]
Humm I wonder if you have the correct permissions for your user? My node-red process is running under the user 'nodered' and that user is in the 'gpio' group. Try that first.

**Comment by Tomadoggy** [comment]
I am using other nodes in node red that are successfully accessing the GPIO pins, so believe that NR has the proper access.  That said however here is what I see for my running NR  process.  From my look it appears as though line #4 shows the rpi-433-v3 is running successfully and has access to the 'gpio' group.

![nrprocesses](https://user-images.githubusercontent.com/62314433/103446730-35d98d80-4c40-11eb-9d9d-a9588640806c.png)

**Comment by jmc265** [comment]
Ok, and I guess running the RFSniffer directly as `/home/pi/.node-red/node_modules/rpi-433-v3/build/RFSniffer 0 27` doesn't result in any output?

**Comment by Tomadoggy** [comment]
That is correct...running the above from a terminal shows nothing...

However running `/home/pi/rpi-rf/scripts/rpi-rf_receive`  does show the captured codes presented

**Comment by jmc265** [comment]
I am not sure then I am afraid. I think you are going to have to do some good old fashioned debugging of the RFSniffer.cpp file. I would advise putting a bunch of `printf` statements in the code then recompiling by running `make` in the src directory. This should re-compile the program to the build directory which you can run again. Hopefully it will output your updated printf statements and you can see the flow of the code and see where it might be stopping.
You could also try using the original version of this code: https://github.com/eroak/rpi-433/ or asking for help on that repo as the original author of this code might reply.

**Comment by Tomadoggy** [comment]
I had to take a break to work on something else, but would like to see if you could offer some additional guidance please.  I am quite new to this but after looking through the .cpp file I am thinking that the following #includes are simply missing from the src directory.  Within my `/home/pi/.node-red/node_modules/rpi-433-v3/src` I do NOT have:

stdlib.h
stdio.h
unistd.h

And am assuming that since they are referenced via the #include lines that they simply are not being found.  Any ideas??

**Comment by Tomadoggy** [comment]
Additional information:

If I go to `https://github.com/eroak/rpi-433/blob/master/src/RFSniffer.cpp` and use this .cpp file and build using the make command...I DO get codes and pulselength captures outputs!  However using your .cpp file at `https://github.com/jmc265/rpi-433-v3/blob/master/src/RFSniffer.cpp`  there is no output.   Does this offer some direction?  

When leaving the working .cpp file active, there is no pass of data over to node red just so you know that I did try that.

**Comment by jmc265** [comment]
I am not sure why the two would be different. The only way to really debug this is by editing the v3 source code, recompiling it and seeing if it works. I would slowly work my way back bit by bit from the v3 code to the v1 code until it starts working, then you will have found the culprit. It is very hard for me to provide more help us I am unable to reproduce the error on my side.

**Comment by Tomadoggy** [comment]
OK...thank you.  Just so I am working from what you have that is functional...are you using the .cpp file without change from:  `https://github.com/jmc265/rpi-433-v3/blob/master/src/RFSniffer.cpp`  ?

When I look at the two versions...there are substantial differences.

If what you have that is working is in any way different...could you possibly post here so I may try that and work from there?

**Comment by Tomadoggy** [comment]
I am going back to the very start of all the installs and configurations to see if I missed anything.  One item jumps out which is your 'Example" code at `https://github.com/jmc265/rpi-433-v3#example`  What exactly am I do be doing with this?  Looks like I have missed a piece of the solution.  Thank you.

**Comment by jmc265** [comment]
That is an example on how to use the library. If you 'npm install' the
library then you can create a file called index.js with that content in it,
then run the example with 'node index.js'.

**Comment by Tomadoggy** [comment]
OK not sure I understand, however from your reply does not look like a requirement for the base install. 

Were you able to verify that the .cpp you have in your v3 install is what you have functional?   As previously stated, I am seeing differences between what you have posted and the `https://github.com/eroak/rpi-433/blob/master/src/RFSniffer.cpp`  file.  As I am very unexperienced in what I am doing here, just trying to figure out how to move from the RFSniffer.cpp towards yours at `https://github.com/jmc265/rpi-433-v3/blob/master/src/RFSniffer.cpp` .

To be clear...what I have done is overwrite the RFSniffer.cpp that is located at `/home/pi/.node-red/node_modules/rpi-433-v3/src` with the code from `https://github.com/eroak/rpi-433/blob/master/src/RFSniffer.cpp`  done a make, then run `/home/pi/.node-red/node_modules/rpi-433-v3/build/RFSniffer` and can successfully see the codes.  From there I am assuming that the node red node should be able to see the same codes, but the debug node on the sniffer still outputs nada...

**Comment by Tomadoggy** [comment]
I understand that this may feel like a dead end to you, but I humbly request a bit more of your time?

I did a bit of a restart on the installs etc. and here is where I am at.  Using everything without change from your v3 install including the unchanged .cpp file I can now receive codes on my Pi with my receiver connected to Pi GPIO #27 but interestingly the only way to see the codes is by running `/home/pi/.node-red/node_modules/rpi-433-v3/build/RFSniffer 0 2` .   If I use options 0 and 27, there is no output.  OK so being encouraged I now go to node red and configure rpi-433-sniffer to listen on pin 2 in its configuration, but still no output on the debug node attached to that node  :(

From there I configured the node red sniffer with every single pin number and re-deployed, but in no instance did I get any codes on the debug.  Does this give you any insights that may help me achieve getting the sniffed codes into node red?
